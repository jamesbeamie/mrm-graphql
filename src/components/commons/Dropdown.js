import React from 'react';
import PropTypes from 'prop-types';
import '../../assets/styles/dropdown.scss';

/**
 * Builds reusable component for Dropdowns
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
class Dropdown extends React.Component {
  state = {
    isVisible: false,
  };

  componentDidUpdate() {
    if (this.state.isVisible) {
      this.dropdownMenu.current.focus();
    }
  }

  dropdownMenu = React.createRef();

  /**
   * It toggles the visibility of the Dropdown
   * @param {object} props
   * @returns {void}
   */
  toggleVisibility = (event) => {
    if (event.type === 'blur' && event.relatedTarget !== null
      && event.target.contains(event.relatedTarget)) {
      return;
    }
    event.preventDefault();
    this.setState({ isVisible: !this.state.isVisible });
  };

  render() {
    return (
      <React.Fragment>
        <div className="dropdown-caret">
          <button onMouseDown={this.toggleVisibility}>
            {this.props.icon ? this.props.icon : <span> &#x2304; </span>}
          </button>
        </div>
        {this.state.isVisible && (
          <div
            id="dropdown-menu"
            className="dropdown-menu"
            role="button"
            onBlur={this.toggleVisibility}
            tabIndex={0}
            ref={this.dropdownMenu}
          >
            {this.props.content}
          </div>
        )}
      </React.Fragment>
    );
  }
}

Dropdown.propTypes = {
  icon: PropTypes.any, // eslint-disable-line
  content: PropTypes.node.isRequired,
};

export default Dropdown;
