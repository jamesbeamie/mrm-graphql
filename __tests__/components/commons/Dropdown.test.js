import React from 'react';
import { shallow } from 'enzyme';
import Dropdown from '../../../src/components/commons/Dropdown';

describe('Sort Component', () => {
  const props = {
    content: `
      <div>
      <span>Rooms</span>
      <span>Wings</span>
      </div>
    `,
  };

  const event = {
    preventDefault: jest.fn(),
    type: 'blur',
    target: {
      contains: jest.fn(),
      relatedTarget: '',
    },
  };
  const wrapper = shallow(<Dropdown {...props} />);

  it('should render component without errors', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should only display button without content', () => {
    expect(wrapper.find('.dropdown-menu').length).toBe(0);
    expect(wrapper.state().isVisible).toBe(false);
  });

  it('should show content when the toggle icon is clicked', () => {
    jest.spyOn(Dropdown.prototype, 'componentDidUpdate');
    wrapper.setState({ isVisible: false });
    wrapper.find('.dropdown-caret button').simulate('mousedown', event);
    expect(Dropdown.prototype.componentDidUpdate.mock.calls.length).toBe(1);
    expect(wrapper.find('.dropdown-menu').length).toBe(1);
    expect(wrapper.state().isVisible).toBe(true);
  });
  it('should not show content when there is a click outside the drop down', () => {
    const e = {
      preventDefault: jest.fn(),
      type: 'blur',
      target: {
        contains: jest.fn().mockReturnValue(true),
      },
      relatedTarget: 'true',
    };
    wrapper.setState({ isVisible: false });
    wrapper.find('.dropdown-caret button').simulate('mousedown', e);
    expect(wrapper.find('.dropdown-menu').length).toBe(0);
    expect(wrapper.state().isVisible).toBe(false);
  });
});

