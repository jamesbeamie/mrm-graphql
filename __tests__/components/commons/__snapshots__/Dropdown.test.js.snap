// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort Component should render component without errors 1`] = `
<Fragment>
  <div
    className="dropdown-caret"
  >
    <button
      onMouseDown={[Function]}
    >
      <span>
         âŒ„ 
      </span>
    </button>
  </div>
</Fragment>
`;
